### PR TITLE
Support Webpack 4

### PR DESCRIPTION
**Pull request by cletusw**
Hack around https://github.com/webpack/webpack/issues/7057

Fixes #5

### CODE REVIEW

File: example/package.json [hunk 1]
```diff
@@ -12,8 +12,9 @@
   "author": "Eemeli Aro <eemeli@gmail.com>",
   "license": "MIT",
   "dependencies": {
-    "messageformat": "^1.0.2",
+    "messageformat": "^1.1.1",
```
Comment (eemeli) [inline example/package.json]: This should work fine with messageformat 2.0.0.

File: index.js [hunk 2]
```diff
@@ -1,5 +1,7 @@
 var loaderUtils = require('loader-utils');
 var MessageFormat = require('messageformat');
+var JavascriptParser = require('./example/node_modules/webpack/lib/Parser');
+var JavascriptGenerator = require('./example/node_modules/webpack/lib/JavascriptGenerator');
```
Comment (eemeli) [inline index.js]: Depending on the dependencies of the example app is not the right thing to do, at all. In fact the example probably should not even be included in the npm-released build. `webpack` should probably be an optional dependency, and some v4 detector be used to determine whether to use the hack or not.

**Review by eemeli — CHANGES_REQUESTED** [review]

File: example/package.json [hunk 3]
```diff
@@ -4,16 +4,17 @@
   "description": "Small example for messageformat-loader",
   "scripts": {
     "build": "webpack",
-    "postinstall": "cd .. && npm install && npm install --no-save messageformat",
+    "postinstall": "cd .. && npm install && npm install --no-save 'file:./example/node_modules/messageformat' 'file:./example/node_modules/webpack'",
```
Comment (cletusw) [inline example/package.json:7]: Uses symlinks so we get the version specified below instead of just latest necessarily.

File: index.js [hunk 4]
```diff
@@ -1,6 +1,17 @@
 var loaderUtils = require('loader-utils');
 var MessageFormat = require('messageformat');
 
+var JavascriptParser;
+var JavascriptGenerator;
+try {
```
Comment (cletusw) [inline index.js:6]: Allows supporting webpack < 4 at the same time.

**Comment by cletusw** [comment]
@eemeli Updated. Let me know what you think, thanks!

**Comment by eemeli** [comment]
I'm growing less convinced that we should hack around Webpack like this. The right thing to do would be to actually submit the PR requested in the upstream issue, and then use that API.

On a sidenote, the `postinstall` hackyness is getting a bit silly. It might make sense to finally bite the bullet and set up a multi-package repo for messageformat using Lerna.

@cletusw Btw, could you sign the JS Foundation CLA?

**Comment by cletusw** [comment]
Signed.

And while I agree that both changes are getting pretty hacky, I'm not in a position right now to do either of the larger changes. I think this change is better than the current state where users are required to know to put "javascript/auto" in their config. Plus, since I've clearly marked off the hacky area, it should be easy to move to an actual published API whenever someone gets around to it.

**Comment by cletusw** [comment]
@eemeli Any thoughts on getting this out there to unblock?

**Comment by eemeli** [comment]
Started looking into Webpack for this a bit for a proper fix, but didn't get deep enough to really figure out how it's all put together before I got distracted; I've spent most of my open-source time on [`yaml`](https://eemeli.org/yaml/) lately.

I'd prefer to leave out the hack, and just ask that users include the `type: 'javascript/auto'` in their config for now. It's pretty certain this'll get fixed in Webpack at some point; until then I think we can just deal with it. :/

**Pull request closed without merging**